File: backend/src/services/geminiService.js
const { genAI } = require('../utils/geminiClient');

/**
 * Interact with Gemini model to generate a response based on context and query
 * @param {string} query - The user's prompt or question
 * @param {string} context - The retrieved context to ground the response
 * @returns {Promise<string>} - The AI generated answer
 */
const generateResponse = async (query, context) => {
    try {
        const model = genAI.getGenerativeModel({ model: "gemini-2.5-flash" }); // or gemini-1.5-flash / gemini-1.5-pro
        
        const systemPrompt = "You are YUDO, a personal AI assistant with access to user's private memory. Answer using the provided context only.";
        
        const fullPrompt = `${systemPrompt}\n\nContext:\n${context}\n\nUser Query: ${query}`;
        
        const result = await model.generateContent(fullPrompt);
        const response = await result.response;
        return response.text();
    } catch (error) {
        console.error("Error generating Gemini response:", error);
        throw error;
    }
};

/**
 * Summarize or process notes based on a query
 * @param {string} query - The user's action query
 * @param {string} notesContext - The notes to process
 * @returns {Promise<string>}
 */
const processNotes = async (query, notesContext) => {
    try {
        const model = genAI.getGenerativeModel({ model: "gemini-2.5-flash" });
        
        const prompt = `You are YUDO, a personal AI assistant. 
Perform the user's requested action (summarization, reasoning, or math calculations) based purely on the provided notes context.

Notes:
${notesContext}

User Query: ${query}

Answer clearly and concisely.`;
        
        const result = await model.generateContent(prompt);
        const response = await result.response;
        return response.text();
    } catch (error) {
        console.error("Error processing notes with Gemini:", error);
        throw error;
    }
};

module.exports = {
    generateResponse,
    processNotes
};

File: backend/services/gemini_service.py
import os
from groq import Groq
from dotenv import load_dotenv
from fastembed import TextEmbedding

load_dotenv()

# Initialize FastEmbed for Embeddings (Runs locally, 768 dims, replaces Gemini)
# We use bge-base-en-v1.5 to match the 768 dimensionality previously used by Gemini
embedding_model = TextEmbedding(model_name="BAAI/bge-base-en-v1.5")

# Initialize Groq for Generation
GROQ_API_KEY = os.getenv("GROQ_API_KEY")
if not GROQ_API_KEY:
    raise ValueError("GROQ_API_KEY must be set in .env")

groq_client = Groq(api_key=GROQ_API_KEY)


def generate_embedding(text: str) -> list[float]:
    """
    Generate a 768-dimension embedding using local FastEmbed model.
    Replaces Gemini because of rate limit / credit issues.
    """
    embeddings = list(embedding_model.embed([text]))
    # Convert numpy array to list of floats
    return embeddings[0].tolist()


def generate_response(query: str, context: str) -> str:
    """
    Generate a RAG-grounded response using Groq (llama-3.3-70b-versatile).
    """
    system_prompt = (
        "You are YUDO, an advanced personal AI assistant. "
        "You have access to the user's private memory/context. "
        "If the context contains relevant information, use it to answer. "
        "If the context is empty or irrelevant to the query, answer the query as a helpful, friendly AI assistant."
    )
    
    context_section = f"Context:\n{context}\n\n" if context.strip() else ""
    full_prompt = f"{system_prompt}\n\n{context_section}User Query: {query}"

    response = groq_client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[
            {"role": "user", "content": full_prompt}
        ]
    )
    return response.choices[0].message.content


def process_notes(query: str, notes_context: str) -> str:
    """
    Perform summarization, reasoning, or math based on user notes using Groq.
    """
    prompt = (
        f"You are YUDO, a personal AI assistant.\n"
        f"Perform the user's requested action (summarization, reasoning, or math) based purely on the notes.\n\n"
        f"Notes:\n{notes_context}\n\nUser Query: {query}\n\nAnswer clearly and concisely."
    )
    
    response = groq_client.chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[
            {"role": "user", "content": prompt}
        ]
    )
    return response.choices[0].message.content
